kubernetes-client/src/main/java/io/fabric8/kubernetes/client/utils/InputStreamPumper.java

```diff
@@ -15,76 +15,86 @@
+  private static final Logger LOGGER = LoggerFactory.getLogger(InputStreamPumper.class);
 
-    final InputStream in;
-    final Callback<byte[]> callback;
-    final Runnable onClose;
-    final AtomicBoolean closed = new AtomicBoolean(false);
+  public interface Writable {
 
-    volatile boolean keepReading = true;
-    Thread thread;
+    void write(byte[] b, int off, int len) throws IOException;
 
-    public InputStreamPumper(InputStream in, Callback<byte[]> callback) {
-      this(in, callback, null);
-    }
+  }
 
-    public InputStreamPumper(InputStream in, Callback<byte[]> callback, Runnable onClose) {
-        this.in = in;
-        this.callback = callback;
-        this.onClose = onClose;
-    }
+  /**
+   * Relies on {@link InputStream#available()} and a Thread sleep to ensure that the reads are interruptible.
+   */
+  public static InputStream asInterruptible(InputStream is) {
+    return new InputStream() {
 
-    @Override
-    public void run() {
-        synchronized (this) {
-          thread = Thread.currentThread();
-        }
-        byte[] buffer = new byte[1024];
-        try {
-          int length;
-          while (keepReading && !Thread.currentThread().isInterrupted() && (length = in.read(buffer)) != -1) {
-                byte[] actual = new byte[length];
-                System.arraycopy(buffer, 0, actual, 0, length);
-                callback.call(actual);
-              }
-        } catch (IOException e) {
-          if (!keepReading) {
-            return;
+      @Override
+      public int read() throws IOException {
+        throw new UnsupportedOperationException();
+      }
+      
+      @Override
+      public int read(byte[] b, int off, int len) throws IOException {
+        while (!Thread.currentThread().isInterrupted()) {
+          if (is.available() > 0) {
+            return is.read(b, off, len);
           }
-          if (!thread.isInterrupted()) {
-            LOGGER.error("Error while pumping stream.", e);
-          } else {
-            LOGGER.debug("Interrupted while pumping stream.");
+          try {
+            // The default sleep interval of 50 milliseconds came from the previous
+            // incarnation of this code.  If this needs tweaked, it can be changed to a parameter.
+            Thread.sleep(50);
+          } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            throw new IOException(e);
           }
-        } finally {
-          close();
         }
-    }
+        throw new IOException();
+      }
+    };
+  }
 
-    public synchronized void close() {
-        keepReading = false;
-        if (thread != null && !thread.isInterrupted()) {
-            thread.interrupt();
-        }
+  /**
+   * See InputStream.transferTo(java.io.OutputStream) in Java 9 or later
+   */
+  public static void transferTo(InputStream in, Writable out) throws IOException {
+    byte[] buffer = new byte[8192];
```

Comment:
I understand the buffer size was determined in a similar way to (https://github.com/fabric8io/kubernetes-client/pull/3169/files#r651965578) and that same considerations apply

Reply:
OK, it was taken from Java's default https://github.com/AdoptOpenJDK/openjdk-jdk11/blob/19fb8f93c59dfd791f62d41f332db9e306bc1422/src/java.base/share/classes/java/io/InputStream.java#L56